### PR TITLE
[new feature] GE2/1 compatibility

### PR DESCRIPTION
**Pull request by mexanick**
Implements required changes in the scan utils for GE2/1 compatibility.
Requires https://github.com/cms-gem-daq-project/cmsgemos/pull/287
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

Tested at GE2/1 integration test stand @904
## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: Delete commented lines here

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: Unnecessary white space

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: I would suggest to have these be generic symlinks rather then specific files as you will need to update this over time (based on GE1/1 experience)

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: I think you want `args.gemType` here

File: testConnectivity.py
Comment (jsturdy) [inline testConnectivity.py]: if, as in @bdorney's comment above, we're removing commented lines, here's a candidate

**Comment by bdorney** [comment]
I didn't quite finish my review of this PR; I'll try to do this by 17h00 today.

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: delete line

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py:777]: Seems like this is a debugging statements that are added but no longer needed....?

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: Seems like this is a debugging statements that are added but no longer needed....?

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py:781]: Seems like this is a debugging statements that are added but no longer needed....?

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py]: Seems like this is a debugging statements that are added but no longer needed....?

File: utils/scanUtils.py
Comment (bdorney) [inline utils/scanUtils.py:20]: In many places you are using `vfatsPerGemVariant[args.gemType]` but here we are using the class data member.  Is there a reason for this inconsistency? e.g. are there some use cases where the class doesn't exist and so it would not be appropriate to take from this after initialization of a `HwVFAT` or `HwOptohybrid` object?  

Otherwise line 19 could probably be put in every script modified above and it might improve readability.  

I do not feel too strongly about this though.

**Comment by bdorney** [comment]
Max array size still at 3072; if there is a corresponding change in `xhal` and `ctp7_modules` to make this array size smaller and an `assert` statement this could cause some errors at runtime no?

https://github.com/mexanick/vfatqc-python-scripts/blob/c8c5a67d712dbf402e6d2a0eca13adf9fcb29d0f/trimChamberV3.py#L168-L170

**Comment by bdorney** [comment]
Also some function documentation and also help menu comments give info on `vfatMask` being a 24 bit number.  This should probably be changed to "at most a 24 bit number" or "either a 12 or 24 bit number depending on gemType" or similar.

Need `vfatsPerGemVariant[gemType]` in the following locations:

https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/testConnectivity.py#L172
https://github.com/mexanick/vfatqc-python-scripts/blob/c8c5a67d712dbf402e6d2a0eca13adf9fcb29d0f/monitorTemperatures.py#L86-L87
https://github.com/mexanick/vfatqc-python-scripts/blob/c8c5a67d712dbf402e6d2a0eca13adf9fcb29d0f/monitorTemperatures.py#L119-L120
https://github.com/mexanick/vfatqc-python-scripts/blob/c8c5a67d712dbf402e6d2a0eca13adf9fcb29d0f/monitorTemperatures.py#L159-L160
https://github.com/mexanick/vfatqc-python-scripts/blob/c8c5a67d712dbf402e6d2a0eca13adf9fcb29d0f/getCalInfoFromDB.py#L23-L24
https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/checkSbitMappingAndRate.py#L136
https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/checkSbitMappingAndRate.py#L241
https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/utils/confUtils.py#L355

Additionally it seems `run_scans.py` has been overlooked:

https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/run_scans.py

e.g. there should be a `parent_parser` that provides the `gemType` and `detType` arguments to all sub commands and then the functions for each sub command should transfer these arguments to the relative command.

The following functions need arguments for `gemType` and `detType`:

https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/utils/phaseUtils.py#L82

Need the `gemType` and `detType` options for the following scripts:

- https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/getCalInfoFromDB.py
- https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/monitorTemperatures.py
- https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/dacScanV3.py
- https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/confChamber.py
- https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/confAllChambers.py
- https://github.com/mexanick/vfatqc-python-scripts/blob/feature/ge21-compatibility/checkSbitMappingAndRate.py

The best way to resolve this if probably to have a `parent_parser` defined with `argparse.ArgumentParser` in `utils/qcoptions.py` and import this `parent_parser` (and any others where appropriate) to the tools that use `ArgumentParser` for their options menus.  Tools still using `optparser` will of course need a manual command line option added (or be migrated to `argparse`).

**Review by bdorney — DISMISSED** [review]
still some incompatibility issues as described here; `run_scans.py` has been overlooked.  Change GE2/1 GBT config files to "generic" symlink rather than hard coded filename/path.

**Comment by mexanick** [comment]
Thanks for the comments, indeed, few scripts has been overlooked. Working on them now

**Comment by mexanick** [comment]
> still some incompatibility issues as described here; `run_scans.py` has been overlooked. Change GE2/1 GBT config files to "generic" symlink rather than hard coded filename/path.

I have 
```python
     if args.gemType == "ge11":
         gbtConfigPath = "{0}/OHv3c/".format(os.getenv("GBT_SETTINGS")) # Ideally this would be a DB read...
     elif args.gemType == "ge21":
         gbtConfigPath = "{0}/OHGE21/".format(os.getenv("GBT_SETTINGS"))
```

Can you clarify what do you mean under "generic" then?

File: utils/scanUtils.py
Comment (mexanick) [reply utils/scanUtils.py:20]: I've tried to use the `vfatsPerGemVariant[args.gemType]` where `HwVFAT` object does not exist, but might have overlooked in some places

**Comment by bdorney** [comment]
> > still some incompatibility issues as described here; `run_scans.py` has been overlooked. Change GE2/1 GBT config files to "generic" symlink rather than hard coded filename/path.
> 
> I have
> 
> ```python
>      if args.gemType == "ge11":
>          gbtConfigPath = "{0}/OHv3c/".format(os.getenv("GBT_SETTINGS")) # Ideally this would be a DB read...
>      elif args.gemType == "ge21":
>          gbtConfigPath = "{0}/OHGE21/".format(os.getenv("GBT_SETTINGS"))
> ```
> 
> Can you clarify what do you mean under "generic" then?

so [here](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265#discussion_r297094218) you are referencing an exit filename; but what I'm using:

```bash
$ ll $GBT_SETTINGS/OHv3c/
total 24K
drwxr-sr--. 2 gemuser gemuser 4.0K Apr 10 15:04 20180717
lrwxrwxrwx. 1 gemuser gemuser   56 May 28 09:56 GBTX_OHv3c_GBT_1.txt -> 20180314/GBTX_OHv3c_GBT_1__2018-03-14_FINAL-REG35-62.txt
lrwxrwxrwx. 1 gemuser gemuser   56 May 28 09:56 GBTX_OHv3c_GBT_2.txt -> 20180314/GBTX_OHv3c_GBT_2__2018-03-14_FINAL-REG35-62.txt
drwxr-sr--. 2 gemuser gemuser 4.0K Jun 14 19:17 20180314
drwxr-sr--. 2 gemuser gemuser 4.0K Jun 14 19:17 20190508
lrwxrwxrwx. 1 gemuser gemuser   56 Jun 17 15:17 GBTX_OHv3c_GBT_0.txt -> 20180314/GBTX_OHv3c_GBT_0__2018-03-14_FINAL-REG35-62.txt
```

So here under the `$ENV` variable I have a set of symlinks, `GBTX_OHv3c_GBT_*.txt` that point to the exact file we need to use.

And in many cases we've had to change the file(s) that those symlinks point too; and in this way such a chance is a *system* change not a codebase change.

This is what I was trying to get at and I think you should use a similar strategy here.

**Review by jsturdy — DISMISSED** [review]
cursory

File: testConnectivity.py
Comment (bdorney) [reply testConnectivity.py:777]: Still outstanding

File: testConnectivity.py
Comment (bdorney) [reply testConnectivity.py:781]: Still outstanding

File: confChamber.py
Comment (bdorney) [inline confChamber.py:53]: Defaulting this to `None` should raise an `OHTypeException` based on these lines:

https://github.com/cms-gem-daq-project/cmsgemos/blob/48e2f269dac0947be0101d2a9f0d943752e2ce61/gempython/tools/vfat_user_functions_xhal.py#L21-L22

When you use this at line 61 I expect an error to be thrown.

File: dacScanV3.py
Comment (bdorney) [inline dacScanV3.py:105]: [same comment](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265/files#r300986304)

File: getCalInfoFromDB.py
Comment (bdorney) [inline getCalInfoFromDB.py]: [same comment](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265/files#r300986304)

File: monitorTemperatures.py
Comment (bdorney) [inline monitorTemperatures.py]: [same comment](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265/files#r300986304)

File: run_scans.py
Comment (bdorney) [inline run_scans.py:708]: [same comment](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265/files#r300986304)

File: sbitReadOut.py
Comment (bdorney) [inline sbitReadOut.py]: [same comment](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265/files#r300986304)

File: sbitThreshScanParallel.py
Comment (bdorney) [inline sbitThreshScanParallel.py]: [same comment](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/265/files#r300986304)

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py:61]: It doesn't seem like the argument `gemType` is actually used in this function.  Given that `amcBoard` is being passed as argument the type is already defined as a class member there.

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py:230]: I think line 209 might need to be changed to have default `detType` changed to `"short"`?

File: testConnectivity.py
Comment (bdorney) [inline testConnectivity.py:312]: Following some comments from @jsturdy I would change line 305 to:

```python
import sys
sys.exit(os.EX_USAGE)
```

This will raise a `SystemExit` exception which can be handled later and is necessary for our py2 to py3 migration in the end.

File: trimChamberV3.py
Comment (bdorney) [inline trimChamberV3.py:51]: same as all the others.

File: ultraLatency.py
Comment (bdorney) [inline ultraLatency.py:173]: same

File: ultraScurve.py
Comment (bdorney) [inline ultraScurve.py:122]: same

File: ultraThreshold.py
Comment (bdorney) [inline ultraThreshold.py:109]: same

**Review by bdorney — CHANGES_REQUESTED** [review]
I think the default choice of `args.detType` will raise an `OHTypeException` when initializing an `HwVFAT` object.  Additionally it would have been nicer to see this added as a parent parser to `qcoptions.py` and then the parent parser used everywhere rather than a copy/paste of arguments.  Now sadly we would need to change things *everywhere* rather than one single location.

**Comment by bdorney** [comment]
Needs `rebase`.

**Comment by jsturdy** [comment]
bump @mexanick

**Comment by mexanick** [comment]
Rebased, but needs testing, please do not merge @jsturdy

**Comment by jsturdy** [comment]
If, as a result of the camp, the bifurcation between GE1/1 and GE2/1 SW versions could be resolved, I would consider that a major win

**Comment by lpetre-ulb** [comment]
> If, as a result of the camp, the bifurcation between GE1/1 and GE2/1 SW versions could be resolved, I would consider that a major win

During my GE2/1 GLIB tests, this PR ended-up working very well. However, I reverted to the main branch (`rpc-playground`), because of a usability flaw. Indeed, the `--gemType` and `--detType` must be specified for every command.

* `--gemType` can be retrieved from the register `GEM_AMC.GEM_SYSTEM.RELEASE.GEM_STATION` starting from GEM_AMC release 3.9.1. For newer release, the GEM station should always be checked against the firmware register.
* `--detType` should be retrieved from the `config_chamber` dictionaries by default. I did not find a way to test both long and short GEBs (GE1/1) with `testConnectivity.py`. This is main reason why we are using the main branch and do not test this PR at ULB.

This PR also powers the FIT setup. @sbutalla, are you encountering any other issues?

**Comment by jsturdy** [comment]
> Merged with current tip of central   `0f4af43` 

should have been a `rebase` rather than a `merge`

**Comment by jsturdy** [comment]
> During my GE2/1 GLIB tests, this PR ended-up working very well. However, I reverted to the main branch (`rpc-playground`), because of a usability flaw. Indeed, the `--gemType` and `--detType` must be specified for every command.

As a stop-gap solution that this was foreseen to be, this is acceptable (provided that the requirement that they be provided exists **only** for `GE2/1` operations)

> 
>     * `--gemType` can be retrieved from the register `GEM_AMC.GEM_SYSTEM.RELEASE.GEM_STATION` starting from GEM_AMC release 3.9.1. For newer release, the GEM station should always be checked against the firmware register.
> 
>     * `--detType` should be retrieved from the `config_chamber` dictionaries by default. I did not find a way to test both long and short GEBs (GE1/1) with `testConnectivity.py`. This is main reason why we are using the main branch and do not test this PR at ULB.

Very little effort should be spent trying to reconcile `detType` and `chamber_config`, and any real thought spent on thinking how best to do this long term through the configuration DB (or text proxy of such, @lmoureaux )

**Comment by lmoureaux** [comment]
> Very little effort should be spent trying to reconcile `detType` and `chamber_config`, and any real thought spent on thinking how best to do this long term through the configuration DB (or text proxy of such, @lmoureaux )

Currently, `gemonlinedb` does not know about the composition of the chambers. Pulling this info from the DB would be easy, and it could also be added to the XML files. This is however a second-order priority with respect to having a working GE1/1 setup.

Generally speaking, I think that information about the hardware layout should either be discovered automatically or come from configuration files. Selecting which link/chamber/... to use is a valid use-case for a command line argument, while telling the system which kind of hardware it is dealing with is not. Just like your OS doesn't ask whether your disk is HDD or SSD every time you read or write data :stuck_out_tongue:

**Comment by jsturdy** [comment]
> Generally speaking, I think that information about the hardware layout should either be discovered automatically or come from configuration files. Selecting which link/chamber/... to use is a valid use-case for a command line argument, while telling the system which kind of hardware it is dealing with is not. Just like your OS doesn't ask whether your disk is HDD or SSD every time you read or write data stuck_out_tongue

Generally I agree with this, and prefer that where functionality depends on these parameters, they should come from the HW itself, while for random tracking metadata, it *could* come from the configuration (this avoids doing additional register requests for **every single** operation, even when the behaviour doesn't change between HW variants)

**Comment by mexanick** [comment]
Generally speaking, things are fixed now, apart from the configuration/detector types. But that should be addressed in a different issue I believe...

**Comment by lpetre-ulb** [comment]
> Generally speaking, things are fixed now, apart from the configuration/detector types.

I believe this PR should be the latest bringing new features to be merged in the legacy branch. The work is already done and the features are important for the GE2/1 US test stands. The PR has been tested at CERN and FIT and seem to work fine.

Fixing the configuration/detector type concerns is out of the scope of this PR and probably out of the scope of the `legacy` branch.

What do you think @mexanick, @jsturdy?

**Comment by jsturdy** [comment]
just rebase it on top of either the current target branch, or the new legacy branch (and change the target)

**Comment by mexanick** [comment]
ok, rebased and retargeted. I believe things *should* work OK, not 100% sure in trimming (although if doesn't work, changes should be very simple and straightforward). Plan to test today with GE2/1 installation at the QC7.

**Review by jsturdy — APPROVED** [review]
Standard legacy cursory

**Comment by jsturdy** [comment]
> ok, rebased and retargeted. I believe things _should_ work OK, not 100% sure in trimming (although if doesn't work, changes should be very simple and straightforward). Plan to test today with GE2/1 installation at the QC7.

Did this get tested? If so (and it passed), please merge

**Comment by mexanick** [comment]
In process of testing, will be completed today. In principle, everything is expected to work out of the box except perhaps iterative trimming, but this will be tested today.

File: confChamber.py
Comment (lpetre-ulb) [reply confChamber.py:53]: Isn't this comment still relevant? For example, the behavior in `checkSbitMappingAndRate.py` is different: https://github.com/cms-gem-daq-project/vfatqc-python-scripts/blob/a9637d6c92d4336eeebcd4ac0aa18cdd77606465/checkSbitMappingAndRate.py#L109

File: monitorTemperatures.py
Comment (lpetre-ulb) [inline monitorTemperatures.py:155]: Do the legacy RPC modules and xHAL interface really support a variable number of VFAT? It is not always the case.

File: testConnectivity.py
Comment (lpetre-ulb) [reply testConnectivity.py:61]: Doesn't it makes sense to review this argument as suggested by Brian?

**Review by lpetre-ulb — APPROVED** [review]
Just a few comments (+ some left by Brian). Nothing outstanding which would block this PR.

**Comment by mexanick** [comment]
The majority of tools are working just fine. Some tools which haven't been used by anyone for ages may exhibit weird behavior. I'd merge it and fix any other issues as soon as they arise

**Comment by lpetre-ulb** [comment]
> The majority of tools are working just fine. Some tools which haven't been used by anyone for ages may exhibit weird behavior. I'd merge it and fix any other issues as soon as they arise

Agreed! The few remaining comments are not outstanding. As soon as the merge is done, a new release should be produced and published.

[I would merge it myself, but I'm not sure what's the merge policy. It also looks like it is different between repos.]

**Comment by jsturdy** [comment]
`2.8.0`: there's no way that these changes could fit into a patch release by any stretch of the imagination
or 
`2.7.12`: there's also likely never been any strict adherence to semver since someone made an initial ninja tag, thereby forcing all subsequent tags to follow suit

**Comment by lpetre-ulb** [comment]
Since the legacy branch is named `legacy-2.7`, I would go for `2.7.12`. As you wrote, semver has never been strictly followed...

If you were to use the more semantically correct version `2.8.0`, I would rename the legacy branch at the same time. This is also the lastest new feature that is envisioned for legacy software.

**Pull request merged**
